core/src/main/scala/kafka/network/RequestConvertToJson.scala

```diff
@@ -0,0 +1,360 @@
+      case res: FetchResponse[_] => FetchResponseDataJsonConverter.write(res.data, version)
+      case res: FindCoordinatorResponse => FindCoordinatorResponseDataJsonConverter.write(res.data, version)
+      case res: HeartbeatResponse => HeartbeatResponseDataJsonConverter.write(res.data, version)
+      case res: IncrementalAlterConfigsResponse => IncrementalAlterConfigsResponseDataJsonConverter.write(res.data, version)
+      case res: InitProducerIdResponse => InitProducerIdResponseDataJsonConverter.write(res.data, version)
+      case res: JoinGroupResponse => JoinGroupResponseDataJsonConverter.write(res.data, version)
+      case res: LeaderAndIsrResponse => LeaderAndIsrResponseDataJsonConverter.write(res.data, version)
+      case res: LeaveGroupResponse => LeaveGroupResponseDataJsonConverter.write(res.data, version)
+      case res: ListGroupsResponse => ListGroupsResponseDataJsonConverter.write(res.data, version)
+      case res: ListOffsetResponse => ListOffsetResponseDataJsonConverter.write(res.data, version)
+      case res: ListPartitionReassignmentsResponse => ListPartitionReassignmentsResponseDataJsonConverter.write(res.data, version)
+      case res: MetadataResponse => MetadataResponseDataJsonConverter.write(res.data, version)
+      case res: OffsetCommitResponse => OffsetCommitResponseDataJsonConverter.write(res.data, version)
+      case res: OffsetDeleteResponse => OffsetDeleteResponseDataJsonConverter.write(res.data, version)
+      case res: OffsetFetchResponse => OffsetFetchResponseDataJsonConverter.write(res.data, version)
+      case res: OffsetsForLeaderEpochResponse => offsetsForLeaderEpochResponseNode(res, version)
+      case res: ProduceResponse => produceResponseNode(res, version)
+      case res: RenewDelegationTokenResponse => RenewDelegationTokenResponseDataJsonConverter.write(res.data, version)
+      case res: SaslAuthenticateResponse => SaslAuthenticateResponseDataJsonConverter.write(res.data, version)
+      case res: SaslHandshakeResponse => SaslHandshakeResponseDataJsonConverter.write(res.data, version)
+      case res: StopReplicaResponse => StopReplicaResponseDataJsonConverter.write(res.data, version)
+      case res: SyncGroupResponse => SyncGroupResponseDataJsonConverter.write(res.data, version)
+      case res: TxnOffsetCommitResponse => TxnOffsetCommitResponseDataJsonConverter.write(res.data, version)
+      case res: UpdateFeaturesResponse => UpdateFeaturesResponseDataJsonConverter.write(res.data, version)
+      case res: UpdateMetadataResponse => UpdateMetadataResponseDataJsonConverter.write(res.data, version)
+      case res: WriteTxnMarkersResponse => WriteTxnMarkersResponseDataJsonConverter.write(res.data, version)
+      case res: VoteResponse => VoteResponseDataJsonConverter.write(res.data, version)
+      case _ => throw new AssertionError(String.format("ApiKey %s is not currently handled in `response`, the " +
+        "code should be updated to do so.", response));
+    }
```

Comment:
nit: we can use scala interpolated string instead of `String.format`, unnecessary semicolon.
also, our error message mentions `ApiKey`, but the templated argument is `response`.

Ditto for request.